src/looprestoration.rs

```diff
@@ -601,185 +605,145 @@ unsafe fn boxsum5<BD: BitDepth>(
 }
 
 #[inline(never)]
-unsafe extern "C" fn selfguided_filter<BD: BitDepth>(
-    mut dst: *mut BD::Coef,
-    mut src: *const BD::Pixel,
+fn selfguided_filter<BD: BitDepth>(
+    mut dst: &mut [BD::Coef; 24576],
+    mut src: &[BD::Pixel; 27300],
     _src_stride: ptrdiff_t,
     w: libc::c_int,
     h: libc::c_int,
     n: libc::c_int,
     s: libc::c_uint,
     bd: BD,
 ) {
-    let sgr_one_by_x: libc::c_uint = (if n == 25 {
-        164 as libc::c_int
-    } else {
-        455 as libc::c_int
-    }) as libc::c_uint;
-    let mut sumsq: [int32_t; 26520] = [0; 26520];
-    let mut A: *mut int32_t = sumsq.as_mut_ptr().offset((2 * 390) as isize).offset(3);
-    let mut sum: [BD::Coef; 26520] = [0.as_(); 26520];
-    let mut B: *mut BD::Coef = sum.as_mut_ptr().offset((2 * 390) as isize).offset(3);
+    let sgr_one_by_x = if n == 25 { 164 } else { 455 };
+
+    // Selfguided filter is applied to a maximum stripe height of 64 + 3 pixels
+    // of padding above and below
+    let mut sumsq = [0; 68 /*(64 + 2 + 2)*/ * REST_UNIT_STRIDE];
+    // By inverting A and B after the boxsums, B can be of size coef instead
+    // of int32_t
+    let mut sum = [0.as_::<BD::Coef>(); 68 /*(64 + 2 + 2)*/ * REST_UNIT_STRIDE];
+
     let step = (n == 25) as libc::c_int + 1;
     if n == 25 {
-        boxsum5::<BD>(sumsq.as_mut_ptr(), sum.as_mut_ptr(), src, w + 6, h + 6);
+        // TODO: Update `boxsum5` to take safe arguments.
+        unsafe {
```

Comment:
Can you add safety comments for all `unsafe` blocks?

Reply:
Done.